Query failure on terminated workflow

## Expected Behavior
The query works for terminated workflows.

## Actual Behavior
```
Exception in thread "main" io.temporal.client.WorkflowQueryException: workflowId='7fc3ceac-eeb3-4468-983d-28716b21e858', runId='ea297429-0663-4c07-96db-e89cad466a09', workflowType='GreetingWorkflow'}
	at io.temporal.internal.sync.WorkflowStubImpl.query(WorkflowStubImpl.java:475)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.queryWorkflow(WorkflowInvocationHandler.java:309)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:272)
	at io.temporal.internal.sync.WorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:178)
	at com.sun.proxy.$Proxy4.queryGreeting(Unknown Source)
	at io.temporal.samples.hello.HelloQuery.main(HelloQuery.java:95)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: io.temporal.internal.replay.InternalWorkflowTaskException: Failure handling event 6 of 'EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED' type. IsReplaying=false, PreviousStartedEventId=3, workflowTaskStartedEventId=0, Currently Processing StartedEventId=3
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEvent(WorkflowStateMachines.java:193)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: io.temporal.internal.replay.InternalWorkflowTaskException: Failure handling event 6 of 'EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED' type. IsReplaying=false, PreviousStartedEventId=3, workflowTaskStartedEventId=0, Currently Processing StartedEventId=3

	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleEvent(ReplayWorkflowRunTaskHandler.java:140)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleWorkflowTaskImpl(ReplayWorkflowRunTaskHandler.java:180)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleQueryWorkflowTask(ReplayWorkflowRunTaskHandler.java:254)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleQueryOnlyWorkflowTask(ReplayWorkflowTaskHandler.java:257)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTask(ReplayWorkflowTaskHandler.java:112)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:314)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:280)
	at io.temporal.internal.worker.PollTaskExecutor.lambda$process$0(PollTaskExecutor.java:73)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Unexpected event:event_id: 6
```

## Steps to Reproduce the Problem
Terminate a workflow and run a query against it.

--- a/temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java
@@ -412,6 +412,7 @@ public final class WorkflowStateMachines {
       case EVENT_TYPE_WORKFLOW_EXECUTION_CANCEL_REQUESTED:
         callbacks.cancel(event);
         break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT:
       case UNRECOGNIZED:
         break;
       default:

--- a/temporal-sdk/src/test/java/io/temporal/workflow/TerminatedWorkflowQueryTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/TerminatedWorkflowQueryTest.java
@@ -1,0 +1,89 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.workflow;
+
+import io.temporal.client.WorkflowClientOptions;
+import io.temporal.client.WorkflowFailedException;
+import io.temporal.client.WorkflowOptions;
+import io.temporal.workflow.shared.SDKTestWorkflowRule;
+import io.temporal.workflow.shared.TestActivities.TestActivitiesImpl;
+import io.temporal.workflow.shared.TestActivities.VariousTestActivities;
+import io.temporal.workflow.shared.TestOptions;
+import io.temporal.workflow.shared.TestWorkflows.TestTraceWorkflow;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class TerminatedWorkflowQueryTest {
+
+  private final TestActivitiesImpl activitiesImpl = new TestActivitiesImpl();
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(TestTraceWorkflowImpl.class)
+          .setActivityImplementations(activitiesImpl)
+          .setWorkflowClientOptions(WorkflowClientOptions.newBuilder().build())
+          .build();
+
+  @Test
+  public void testShouldReturnQueryResultAfterWorkflowTimeout() {
+    WorkflowOptions options =
+        TestOptions.newWorkflowOptionsWithTimeouts(testWorkflowRule.getTaskQueue())
+            .toBuilder()
+            .setWorkflowRunTimeout(Duration.ofSeconds(1))
+            .build();
+    TestTraceWorkflow client =
+        testWorkflowRule.getWorkflowClient().newWorkflowStub(TestTraceWorkflow.class, options);
+
+    Assert.assertThrows(
+        "Workflow should throw because of timeout",
+        WorkflowFailedException.class,
+        () -> client.execute(SDKTestWorkflowRule.useExternalService));
+
+    Assert.assertEquals(1, client.getTrace().size());
+    Assert.assertEquals("started", client.getTrace().get(0));
+  }
+
+  public static class TestTraceWorkflowImpl implements TestTraceWorkflow {
+
+    private final List<String> trace = new ArrayList<>();
+
+    @Override
+    public String execute(boolean useExternalService) {
+      VariousTestActivities localActivities =
+          Workflow.newLocalActivityStub(
+              VariousTestActivities.class, TestOptions.newLocalActivityOptions());
+
+      trace.add("started");
+      localActivities.sleepActivity(5000, 123);
+      trace.add("finished");
+      return "";
+    }
+
+    @Override
+    public List<String> getTrace() {
+      return trace;
+    }
+  }
+}

